feature/1024: Implement and standardize Cart API endpoints

irvaniamirali opened the pull request:
Implemented all Cart endpoints (`POST /cart/`, `GET /cart/{user_id}`, `DELETE /cart/{user_id}`, `POST /cart/items`, `DELETE /cart/items/{user_id}/{product_id}`) using `cart/usecases` and `CartItemRepository`. Includes proper validation, error handling, and RESTful standards. Resolves #24.

irvaniamirali committed:
chore(cart): include Cart router in main FastAPI app

irvaniamirali committed:
refactor(deps): migrate db_session injection to PEP-593 Annotated style

irvaniamirali committed:
feat(cart): implement CartItemRepository with CRUD operations

irvaniamirali committed:
refactor(cart): define CartItemRepositoryInterface for consistent repository abstraction

irvaniamirali committed:
feat(cart): implement and standardize all Cart endpoints

irvaniamirali committed:
feat(cart): implement AddItemToCart use case for single item addition

irvaniamirali committed:
feat(cart): implement CreateCartFromItems use case for bulk cart creation

irvaniamirali committed:
feat(cart): implement ReadCart use case for retrieving user cart contents

irvaniamirali committed:
feat(cart): implement RemoveItemFromCart use case for deleting single cart item

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            return cart_item
+        except SQLAlchemyError as e:
+            await self.session.rollback()
+            raise DatabaseOperationException(
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            return cart_items
+        except SQLAlchemyError as e:
+            await self.session.rollback()
+            raise DatabaseOperationException(
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            stmt = select(self.model_class).where(self.model_class.user_id == user_id)
+            result = await self.session.execute(stmt)
+            return result.scalars().all()
+        except SQLAlchemyError as e:
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+        try:
+            await self.session.execute(CartItem.__table__.delete().where(CartItem.user_id == user_id))
+            await self.session.commit()
+        except SQLAlchemyError as e:
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            await self.session.commit()
+        except SQLAlchemyError as e:
+            await self.session.rollback()
+            raise DatabaseOperationException(operation="delete", message=str(e), data={"user_id": user_id})
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            result = await self.session.execute(stmt)
+            return result.scalar_one_or_none()
+        except SQLAlchemyError as e:
+            raise DatabaseOperationException(
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            result = await self.session.execute(stmt)
+            cart_item = result.scalar_one_or_none()
+
+            if not cart_item:
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan commented on app/modules/cart/repository.py:
```diff
@@ -1,22 +1,95 @@
+            await self.session.delete(cart_item)
+            await self.session.commit()
+
+        except SQLAlchemyError as e:
```
we usually handle DB exceptions at a higher layer (e.g., service layer/usecase) rather than the repository itself, to keep our repositories focused on data access logic and avoid coupling them with application-specific error handling.

babakjahan reviewed (changes requested):
Consider letting exceptions propagate and catching them in the usecase/business layer

irvaniamirali committed:
refactor(usecase): move repository exception handling to RemoveItemFromCart

irvaniamirali committed:
refactor(usecase): move repository exception handling to ReadCart

irvaniamirali committed:
refactor(usecase): move repository exception handling to CreateCartFromItems

irvaniamirali committed:
refactor(usecase): move repository exception handling to AddItemToCart

irvaniamirali committed:
refactor(cart): move repository exception handling to usecases

- Removed direct SQLAlchemy exception handling from CartItemRepository.
- All database-related exceptions (IntegrityError, SQLAlchemyError, etc.)
  are now propagated to the corresponding usecases.
- Usecases handle exceptions consistently with DatabaseOperationException,
  DuplicateEntryException, and EntityNotFoundException.
- Keeps repository focused on CRUD operations (SRP) and aligns with SOLID principles.
